# Local Analysis for Anomaly Detection within The Kafka-DataStreams

Let's import some useful libraries

In [1]:
import csv
import pandas as pd
from pandas import DataFrame
import seaborn as sns

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


Let's move to the data Directory 

In [2]:
cd /content/drive/My Drive/

/content/drive/My Drive


Let's link our data files

In [0]:
df = pd.read_json('clickStream.json', lines=True)
df2 = pd.read_json('displayStream.json', lines=True)

In [4]:
df.head(5)

eventType  ...                          impressionId
0     click  ...  b327f163-f47f-4693-88a6-817e33977c78
1     click  ...  cf821c1c-5735-4c0a-914f-766f8372f11c
2     click  ...  6167acae-c894-4918-b013-031b6589d8d7
3     click  ...  fe7c3093-640c-4a8b-b815-0f5a53ff5a36
4     click  ...  2f9547ca-7059-4a65-80de-6631b2252805

[5 rows x 5 columns]

In [24]:
df2.head(5)

eventType  ...                          impressionId
0   display  ...  426cd1c9-e299-4b91-8ead-769b58d6b300
1   display  ...  0e124861-ecdf-494d-bbcc-069a13ebe2a6
2   display  ...  aca5e651-2a63-4fc9-9322-5aa2c5f17b86
3   display  ...  d8e191b9-88d2-4e03-8f89-61dc9a382ab0
4   display  ...  96b812dc-00ee-4af3-a851-524eb995f4a5

[5 rows x 5 columns]

## Let's detect the first patern

In [0]:
pd_clickStream = df
pd_clickStream = pd_clickStream.drop_duplicates( subset = ["uid"], keep ="first")
pd_clickStream2 = pd_clickStream.set_index('uid')
pd_clickStream3 = pd_clickStream2['ip'].unique()

In [29]:
len(pd_clickStream2)

1577

In [30]:
len(pd_clickStream3)

731

As there are 731 unique IP within 1577 uid's rows in the clickStream, some IP's are repeated. 

In [0]:
pd_displayStream = df
pd_displayStream = pd_displayStream.drop_duplicates( subset = ["uid"], keep ="first")
pd_displayStream2 = pd_displayStream.set_index('uid')
pd_displayStream3 = pd_displayStream2['ip'].unique()

In [32]:
len(pd_displayStream2)

1577

In [33]:
len(pd_displayStream3)

731

Likewise, as there are 731 unique IP within 1577 uid's rows in the displayStream, some IP's are repeated. \\
**THIS IS LIKELY TO BE A FRAUD!!**

# Let's detect the second patern

In [34]:
pd_clickStream = df
pd_clickStream = pd_clickStream.drop_duplicates( subset = ["ip"], keep ="first")
pd_clickStream2 = pd_clickStream.set_index('ip')
pd_clickStream3 = pd_clickStream2['uid'].unique()
len(pd_displayStream2)

1577

In [35]:
len(pd_displayStream3)

731

There are some users who change their ips while keeping the same uid in the clickStream.

In [36]:
pd_displayStream = df
pd_displayStream = pd_displayStream.drop_duplicates( subset = ["ip"], keep ="first")
pd_displayStream2 = pd_displayStream.set_index('ip')
pd_displayStream3 = pd_displayStream2['uid'].unique()
len(pd_displayStream2)

2628

In [37]:
len(pd_displayStream3)

654

There are some users who change their ips while keeping the same uid in the displayStream. \\
**THIS IS LIKELY TO BE OUR SECOND FRAUD!!**

## To sum up, we detected 2 Patterns. 
## Now, let's move forward to flink implementation! 